### PR TITLE
Port call-process-region

### PR DESCRIPTION
**Pull request by agraven**
This solves #1229.

Some of the c functions used were ported as well. Additionally, since I ported `record_unwind_protect_int` I ported all the other `record_unwind_protect` function while I was at it since they are almost identical.

### CODE REVIEW

File: rust_src/src/callproc.rs [hunk 1]
```diff
@@ -79,4 +83,53 @@ pub fn call_process_lisp(args: &mut [LispObject]) -> LispObject {
     })
 }
 
+#[lisp_fn(min = "3")]
```
Comment (shaleh) [inline rust_src/src/callproc.rs]: The docstring is missing.

File: rust_src/src/callproc.rs [hunk 2]
```diff
@@ -79,4 +83,53 @@ pub fn call_process_lisp(args: &mut [LispObject]) -> LispObject {
...
+pub fn call_process_region(args: &mut [LispObject]) -> LispObject {
+    let mut start = args[0];
+    let mut end = args[1];
+    let mut infile = Qnil;
+    let spec = c_specpdl_index();
+
+    let empty_input = if start.is_string() {
+        LispStringRef::from(start).len_chars() == 0
+    } else if start.is_nil() {
+        let buffer = &ThreadState::current_buffer_unchecked();
```
Comment (shaleh) [inline rust_src/src/callproc.rs]: Why the reference?

Comment (agraven) [reply rust_src/src/callproc.rs]: That was a mistake, I have a habit of forgetting the behaviour of `ExternalPtr`.

File: rust_src/src/callproc.rs (hunk 1 above)
Comment (agraven) [reply rust_src/src/callproc.rs]: Fixed

File: rust_src/src/callproc.rs [hunk 3]
```diff
@@ -79,4 +83,83 @@ pub fn call_process_lisp(args: &mut [LispObject]) -> LispObject {
...
+///
+/// usage: (call-process-region START END PROGRAM &optional DELETE BUFFER DISPLAY &rest ARGS)
+#[lisp_fn(min = "3")]
+pub fn call_process_region(args: &mut [LispObject]) -> LispObject {
+    let mut start = args[0];
+    let mut end = args[1];
+    let mut infile = Qnil;
+    let spec = c_specpdl_index();
+
+    let empty_input = if start.is_string() {
```
Comment (nickdrozd) [inline rust_src/src/callproc.rs]:
```suggestion
    let empty_input = if let start.as_string() = Some(string) {
        string.len_chars() == 0
```

File: rust_src/src/eval.rs [hunk 4]
```diff
@@ -37,6 +41,58 @@ use crate::{
  *   and temporaries from garbage collection while it needs them.      *
  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
+#[no_mangle]
```
Comment (nickdrozd) [inline rust_src/src/eval.rs]: Could you break these out into a separate commit? It seems like a significant change that's worth recording apart from `call-process-region`.

**Comment by nickdrozd** [comment]
I actually tried and failed to port this two months ago, so great job!

File: rust_src/src/callproc.rs (hunk 3 above)
Comment (shaleh) [reply rust_src/src/callproc.rs]: @nickdrozd has this backwards `if let Some(string) = start.as_string() {`

File: rust_src/src/eval.rs (hunk 4 above)
Comment (agraven) [reply rust_src/src/eval.rs]: Sure thing.

Comment (agraven) [reply rust_src/src/eval.rs]: These changes are now in #1248.

**Comment by brotzeit** [comment]
Can we put the function declaration of `create_temp_file` in src/callproc.h ?

**Comment by agraven** [comment]
@brotzeit Absolutely.

File: src/lisp.h [hunk 5]
```diff
@@ -4387,6 +4387,7 @@ extern void init_callproc (void);
 extern void set_initial_environment (void);
 extern void syms_of_callproc (void);
 extern Lisp_Object call_process (ptrdiff_t, Lisp_Object *, int, ptrdiff_t);
+extern int create_temp_file (ptrdiff_t, Lisp_Object *, Lisp_Object *);
 
 /* Defined in doc.c.  */
```
Comment (brotzeit) [inline src/lisp.h]: Please add a comment like this.
